File: ios/FaceLivenessDetectorModule.swift
import ExpoModulesCore

public class FaceLivenessDetectorModule: Module {
  public func definition() -> ModuleDefinition {
    Name("FaceLivenessDetector")

    AsyncFunction("setAuthCredentials") { (credentials: [String: String]) in
      guard let accessKeyId = credentials["accessKeyId"] else {
        throw FaceLivenessDetectorCredentialStoreError.missingCredentials
      }

      guard let secretAccessKey = credentials["secretAccessKey"] else {
        throw FaceLivenessDetectorCredentialStoreError.missingCredentials
      }

      guard let sessionToken = credentials["sessionToken"] else {
        throw FaceLivenessDetectorCredentialStoreError.missingCredentials
      }

      guard let expiration = credentials["expiration"] else {
        throw FaceLivenessDetectorCredentialStoreError.invalidExpiration
      }

      try FaceLivenessDetectorCredentialStore.set(
        FaceLivenessDetectorAWSTemporaryCredentials(
          accessKeyId: accessKeyId,
          secretAccessKey: secretAccessKey,
          sessionToken: sessionToken,
          expiration: try parseFaceLivenessDetectorExpiration(expiration)
        )
      )
    }

    AsyncFunction("clearAuthCredentials") {
      FaceLivenessDetectorCredentialStore.clear()
    }

    View(FaceLivenessDetectorExpoView.self) {
      Prop("sessionId") { (view: FaceLivenessDetectorExpoView, value: String) in
        view.sessionId = value
      }

      Prop("region") { (view: FaceLivenessDetectorExpoView, value: String) in
        view.region = value
      }

      Prop("disableStartView") { (view: FaceLivenessDetectorExpoView, value: Bool) in
        view.disableStartView = value
      }

      Prop("challengeOptions") { (view: FaceLivenessDetectorExpoView, value: [String: Any]?) in
        view.challengeOptionsPayload = value
      }

      Events("onAnalysisComplete", "onError")
    }
  }
}

File: android/src/main/java/expo/modules/faceliveness/FaceLivenessDetectorModule.kt
package expo.modules.faceliveness

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class FaceLivenessDetectorModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("FaceLivenessDetector")

    AsyncFunction("setAuthCredentials") { credentials: Map<String, String> ->
      FaceLivenessDetectorCredentialStore.set(
        FaceLivenessDetectorAWSTemporaryCredentials(
          accessKeyId = credentials["accessKeyId"]
            ?: throw IllegalArgumentException("accessKeyId is required"),
          secretAccessKey = credentials["secretAccessKey"]
            ?: throw IllegalArgumentException("secretAccessKey is required"),
          sessionToken = credentials["sessionToken"]
            ?: throw IllegalArgumentException("sessionToken is required"),
          expiration = credentials["expiration"]
            ?: throw IllegalArgumentException("expiration is required")
        )
      )
    }

    AsyncFunction("clearAuthCredentials") {
      FaceLivenessDetectorCredentialStore.clear()
    }

    View(FaceLivenessDetectorExpoView::class) {
      Prop("sessionId") { view: FaceLivenessDetectorExpoView, value: String ->
        view.sessionId = value
      }

      Prop("region") { view: FaceLivenessDetectorExpoView, value: String ->
        view.region = value
      }

      Prop("disableStartView") { view: FaceLivenessDetectorExpoView, value: Boolean ->
        view.disableStartView = value
      }

      Prop("challengeOptions") { view: FaceLivenessDetectorExpoView, value: Map<String, Any>? ->
        view.challengeOptions = value
      }

      Events("onAnalysisComplete", "onError")
    }
  }
}
